src/robot/running/context.py

```diff
@@ -126,14 +126,20 @@ def in_teardown(self):
     def variables(self):
         return self.namespace.variables
 
-    @property
-    def continue_on_failure(self):
+    def continue_on_failure(self, default_continue=False):
         parents = ([self.test] if self.test else []) + self.user_keywords
         if not parents:
+            return default_continue
+        if 'robot:stop-on-failure' in parents[-1].tags:
             return False
         if 'robot:continue-on-failure' in parents[-1].tags:
             return True
-        return any('robot:recursive-continue-on-failure' in p.tags for p in parents)
+        for parent in reversed(parents):
+            if 'robot:recursive-stop-on-failure' in parent.tags:
+                return False
+            if 'robot:recursive-continue-on-failure' in parent.tags:
+                return True
```

Comment:
It would be possible to combine checking modes into single place and remove the `if not parents` block at the same time:

```python
first = True
for parent in reversed(parents):
    if ('robot:recursive-continue-on-failure' in parent.tags
            or first and 'robot:continue-on-failure' in parent.tags):
        return True
    if ('robot:recursive-stop-on-failure' in parent.tags
            or first and 'robot:stop-on-failure' in parent.tags):
        return False
    first = False
```

Could also use `enumerate` and check the index instead of using the `first` flag. Adding separate helper methods for checking tags that could be considered as well:

```python
for index, parent in enumerate(reversed(parents)):
    if self._continue_by_tags(parent.tags, index == 0):
        return True
    if self._stop_by_tags(parent.tags, index == 0):
        return False
```

These are just ideas, the current code is good as it is and can be fine-tuned later if really needed.

Reply:
I like the optimization suggestion, took the enumerate one without helper..